Support for struct attributes with reserved names

flash-gordon opened the pull request:
This fixes #37 but slows struct initialization a bit (actually 1.5x difference), see https://gist.github.com/flash-gordon/f9b99f8c28f0a0d65902cae18035a45f for details. I also added more friendly errors on missing or unknown keys.

flash-gordon committed:
Support for struct attributes with reserved names

fixes #37

flash-gordon commented:
@solnic wdyt?

solnic commented on lib/rom/repository/struct_attributes.rb:
```diff
@@ -17,26 +17,27 @@ def initialize(attributes)
+            actual = values.keys
+            unknown, missing = actual - attributes, attributes - actual
+
+            if unknown.any? || missing.any?
```
`size > 0` should be faster

flash-gordon replied:
it should, but it doesn't affect the overall performance

```
Warming up --------------------------------------
              manual    14.248k i/100ms
     manual with any    14.320k i/100ms
Calculating -------------------------------------
              manual    149.134k (± 1.9%) i/s -    755.144k in   5.065287s
     manual with any    148.518k (± 2.4%) i/s -    744.640k in   5.016815s

Comparison:
              manual:   149134.0 i/s
     manual with any:   148518.2 i/s - same-ish: difference falls within error
```

flash-gordon replied:
updated

flash-gordon committed:
Check size instead of using .any?

solnic commented:
👍

flash-gordon committed:
Use .fetch for extracting values from hash

Also check for unknown keys more lazily. This speeds initialization up
by ~3x. :heart: @AMHOL

AMHOL commented:
:+1: Nice one
